crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -120,22 +209,60 @@ impl FromPeers {
     async fn relay_messages_to(
         &mut self,
         outgoing_mqtt: &mut mpsc::UnboundedSender<MqttMessage>,
+        client: mqtt_channel::AsyncClient,
+        subscriptions: Arc<Mutex<TopicFilter>>,
```

Comment:
I would combine these two fields of an `mqtt_chanel::Connection`  in a single one (say a  `subscriber`)  that handles the whole logic below to compute the new subscription set and update `mqtt_channel::AsyncClient`.